File: JINSEOP/백준_혼자_풀기/DP/스티커.cpp
#include <iostream>
using namespace std;

int tc = 0, num = 0;
int scores[2][100001], dp[2][100001];
int main() {
	cin >> tc;
	for (int i = 0; i < tc; i++) {
		cin >> num;
		for (int j = 0; j < num; j++) {
			cin >> scores[0][j];
		}
		for (int j = 0; j < num; j++) {
			cin >> scores[1][j];
		}

		dp[0][0] = scores[0][0]; dp[1][0] = scores[1][0];
		dp[0][1] = scores[0][1] + dp[1][0]; 
		dp[1][1] = scores[1][1] + dp[0][0]; // 미리 설정

		for (int j = 2; j < num; j++) {
			dp[0][j] = max(scores[0][j] + dp[1][j - 1], scores[0][j] + dp[1][j - 2]);
			dp[1][j] = max(scores[1][j] + dp[0][j - 1], scores[1][j] + dp[0][j - 2]);
		}

		cout << max(dp[0][num - 1], dp[1][num - 1]) << '\n';
	}

	return 0;
}

File: JINSEOP/백준 자바로 풀기/DP/스티커.java
import java.io.IOException;
import java.lang.reflect.Array;
import java.util.*;

public class Main {
    public static void clearDp(long[] dp, int num){
        for(int i = 1; i <= num; i++)
            dp[i] = 0;
    }

    public static void main(String[] args) throws IOException {
        Scanner sc = new Scanner(System.in);
        int tc = sc.nextInt();

        for(int i = 0; i < tc; i++){
            int num = sc.nextInt();
            int[][] stickers = new int[2][100000];
            int[][] dp = new int[2][100000];

            for(int row = 0; row < 2; row++)
                for(int col = 0; col < num; col++)
                    stickers[row][col] = sc.nextInt();
            // 스티커 입력받음

            dp[0][0] = stickers[0][0]; dp[1][0] = stickers[1][0]; // basis
            dp[0][1] = dp[1][0] + stickers[0][1]; dp[1][1] = dp[0][0] + stickers[1][1];

            for(int col = 2; col < num; col++){
                dp[0][col] = Math.max(dp[1][col-2] + stickers[0][col], dp[1][col-1] + stickers[0][col]);
                dp[1][col] = Math.max(dp[0][col-2] + stickers[1][col], dp[0][col-1] + stickers[1][col]);
            }

            System.out.println(Math.max(dp[0][num - 1], dp[1][num - 1]));
        }
    }
}
